### PR TITLE
[ci] [release-automation] Add pre-release checks

### PR DESCRIPTION
**Pull request by khluu**
## Changes
- Install git and add a non-root user for `forge_x86_64.Dockerfile` (using root user causes error when calling `git ....`)
- Add `--check` flag to `weekly_green_metric` and raise Exception if number of release blockers is not 0.
- Add a script to call `weekly_green_metric` to check whether number of release blockers is down to 0
- Add a step to verify that `BUILDKITE_COMMIT` is the most recent git commit hash
- Allow trigger step in config
- Add a `pre-release check` test group that includes:
  - Checking how many release blockers are there and don't proceed further unless it's 0
  - Trigger postmerge build from `release-automation` build
  - Trigger `release nightly` test from `release-automation` build

## Checks
- [This build on this PR branch](https://buildkite.com/ray-project/release-automation/builds/166) shows that `check_release_blockers` returns 10 blockers and failed there, which didn't trigger tests from other pipelines.
- [This build](https://buildkite.com/ray-project/release-automation/builds/165) from another identical branch intentionally let `check_release_blockers` pass, thus was able to trigger postmerge & 3 release tests with the right commit & branch sent in as env var. The tests failed since I cancelled them.
-

### CODE REVIEW

**Review by aslonnie — COMMENTED** [review]
I think the pipeline forge group change just got merged. could you rebase/merge master head?

**Comment by khluu** [comment]
@aslonnie I just rebased against master here

File: .buildkite/release-automation/config.yml [hunk 1]
```diff
@@ -13,6 +13,7 @@ runner_queues:
   macos: macos
   macos-arm64: macos-pr-arm64
   windows: windows_queue_pr
+  small_branch: runner_queue_small_branch
```
Comment (aslonnie) [inline .buildkite/release-automation/config.yml:16]: could you split this config change into a separate PR?

File: ci/ray_ci/automation/weekly_green_metric.py [hunk 2]
```diff
@@ -25,7 +25,7 @@ def main(production: bool) -> None:
     init_global_config(bazel_runfile("release/ray_release/configs/oss_config.yaml"))
     blockers = TestStateMachine.get_release_blockers()
     logger.info(f"Found {blockers.totalCount} release blockers")
-
+    print(blockers.totalCount)
```
Comment (aslonnie) [inline ci/ray_ci/automation/weekly_green_metric.py:28]: rather than doing this, can we add an flag/arg `--check` and perform the check inside the script? rather than doing another redirection via bash?

File: .buildkite/release-automation/pre_release.rayci.yml [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+  - label: "Trigger Postmerge test"
+    trigger: "postmerge"
+    if: build.env("COMMIT") != null && build.env("BRANCH") != null
+    build:
+      commit: "${COMMIT}"
+      branch: "${BRANCH}"
+      message: "Triggered by release-automation build #${BUILDKITE_BUILD_NUMBER}"
+      env:
+        NIGHTLY: 1
```
Comment (can-anyscale) [inline .buildkite/release-automation/pre_release.rayci.yml:24]: what 'nightly: 1' does?

also a todo, are the non-flaky mac os tests blocking yet?

Comment (khluu) [reply .buildkite/release-automation/pre_release.rayci.yml:24]: My idea is to add `NIGHTLY=1` as an environment variable in every scheduled nightly build, and then enforce that we only build images and tag `nightly` when `NIGHTLY=1` which only happens on scheduled nightly runs. It's used in this PR: https://github.com/ray-project/ray/pull/43486

Comment (can-anyscale) [reply .buildkite/release-automation/pre_release.rayci.yml:24]: got you

File: ci/ray_ci/automation/weekly_green_metric.py (hunk 2 above)
Comment (khluu) [reply ci/ray_ci/automation/weekly_green_metric.py]: done!

File: .buildkite/release-automation/config.yml (hunk 1 above)
Comment (khluu) [reply .buildkite/release-automation/config.yml]: done! PR is #43519

File: .buildkite/release-automation/pre_release.rayci.yml (hunk 3 above)
Comment (khluu) [reply .buildkite/release-automation/pre_release.rayci.yml:24]: ^ just to clarify, we still build images and push with the commit tag normally, just not have `nightly` alias when `NIGHTLY != 1` or `NIGHTLY` doesn't exist in the build env

File: .buildkite/release-automation/pre_release.rayci.yml [hunk 4]
```diff
@@ -0,0 +1,36 @@
+group: Pre-release checks
+steps:
+  - label: "Check release blockers"
+    key: check-release-blockers
+    depends_on: forge
+    job_env: forge
+    instance_type: small_branch
```
Comment (khluu) [inline .buildkite/release-automation/pre_release.rayci.yml:7]: This PR will be merged after #43519

Comment (khluu) [reply .buildkite/release-automation/pre_release.rayci.yml:7]: to not cause errors for this line

File: ci/ray_ci/automation/weekly_green_metric.py (hunk 2 above)
Comment (khluu) [reply ci/ray_ci/automation/weekly_green_metric.py]: Ran a build on another identical branch (so I could add the branch queue in to test): https://buildkite.com/ray-project/release-automation/builds/172#_
It correctly asserts the release blocker count and fails the step.

File: ci/ray_ci/automation/weekly_green_metric.py [hunk 5]
```diff
@@ -39,6 +47,11 @@ def main(production: bool) -> None:
         )
         logger.info("Weekly green metric updated successfully")
 
+    if check:
+        assert (
+            blockers.totalCount == 0
+        ), "There are still release blockers. Please resolve them before proceeding."
```
[the remark this answers is not in the document]
Comment (khluu) [reply ci/ray_ci/automation/weekly_green_metric.py]: done! I raised `ValueError` instead

File: .buildkite/release-automation/pre_release.rayci.yml [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+  - block: "Trigger tests & image builds"
+    depends_on:
+      - step: check-release-blockers
+        allow_failure: true
+
+  - label: "Trigger Postmerge test"
+    trigger: "postmerge"
+    if: build.env("COMMIT") != null && build.env("BRANCH") != null
+    build:
+      commit: "${COMMIT}"
```
Comment (khluu) [inline .buildkite/release-automation/pre_release.rayci.yml:20]: Instead of having users input their own `COMMIT` env var in the build, I think we can just set the commit & branch when creating build and use `$BUILDKITE_COMMIT` instead here.
That also helps maintain the same commit hash across the whole pipeline build (as `HEAD` is a moving target) which is also used for downloading S3 wheels & as `RAY_HASH` when sanity check the wheels.
For scheduled run of this pipeline, commit is just set to `HEAD` automatically.
What do you think? @aslonnie

Comment (aslonnie) [reply .buildkite/release-automation/pre_release.rayci.yml:20]: yes, you should just use `BUILDKITE_COMMIT` and `BUILDKITE_BRANCH`

Comment (aslonnie) [reply .buildkite/release-automation/pre_release.rayci.yml:20]: make sure `BUILDKITE_COMMIT` is not set to `HEAD` but actually a specific commit.

Comment (khluu) [reply .buildkite/release-automation/pre_release.rayci.yml:20]: Doesn't `BUILDKITE_COMMIT` get updated into an actual commit hash after `init` step, like in [this build](https://buildkite.com/ray-project/release-automation/builds/171#018df17d-7101-4ae1-9f19-2e7db5eb3b43/155-565)? All other steps have `BUILDKITE_COMMIT=<commit_hash>` instead of `HEAD` like `init` step.

File: ci/ray_ci/automation/weekly_green_metric.py [hunk 7]
```diff
@@ -39,6 +47,11 @@ def main(production: bool) -> None:
         )
         logger.info("Weekly green metric updated successfully")
 
+    if check and blockers.totalCount != 0:
+        raise ValueError(
```
Comment (aslonnie) [inline ci/ray_ci/automation/weekly_green_metric.py:51]: could you just raise `Exception`, or maybe just print a line to stderr, and `sys.exit(1)`? `ValueError` means an argument has inappropriate value.

File: .buildkite/release-automation/pre_release.rayci.yml (hunk 6 above)
Comment (aslonnie) [reply .buildkite/release-automation/pre_release.rayci.yml:20]: https://buildkite.com/docs/pipelines/environment-variables#BUILDKITE_COMMIT

Comment (aslonnie) [reply .buildkite/release-automation/pre_release.rayci.yml:20]: you can add a check step to check that the value of `$BUILDKITE_COMMIT` is the same as `$(giv rev-parse HEAD)` and fail otherwise.

Comment (khluu) [reply .buildkite/release-automation/pre_release.rayci.yml]: done!

File: ci/ray_ci/automation/weekly_green_metric.py (hunk 7 above)
Comment (khluu) [reply ci/ray_ci/automation/weekly_green_metric.py]: done!

File: .buildkite/release-automation/pre_release.rayci.yml (hunk 3 above)
Comment (aslonnie) [reply .buildkite/release-automation/pre_release.rayci.yml:30]: let's maybe remove the envs for now. and leave a todo mark here, and we backfill this env thing in another PR?

File: .buildkite/release-automation/pre_release.rayci.yml [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+  - block: "Trigger tests & image builds"
+    depends_on:
+      - step: check-release-blockers
+        allow_failure: true
+
+  - label: "Check commit hash"
+    key: check-commit-hash
+    job_env: forge
+    commands:
+      - if [[ "${BUILDKITE_COMMIT}" != "$(git rev-parse HEAD)" ]]; then echo "Commit hash mismatch"; exit 1; else echo "Commit hash matches"; fi
```
Comment (aslonnie) [inline .buildkite/release-automation/pre_release.rayci.yml:20]: could you move this into a bash script on the repo?

Comment (khluu) [reply .buildkite/release-automation/pre_release.rayci.yml]: Triggered a build and checking commit hash works normally: https://buildkite.com/ray-project/release-automation/builds/184#018df6bb-4fe1-44f0-a9e0-a925dbe05c9f

**Review by can-anyscale — APPROVED** [review]

**Review by aslonnie — APPROVED** [review]
